packages/flutter/lib/src/widgets/text.dart

```diff
@@ -713,3 +721,548 @@ class Text extends StatelessWidget {
     }
   }
 }
+
+// In practice some selectables like widgetspan shift several pixels. So when
+// the vertical position diff is within the threshold, compare the horizontal
+// position to make the compareScreenOrder function more robust.
+const double _kSelectableVerticalComparingThreshold = 3.0;
+
+class _SelectableTextContainerDelegate extends MultiSelectableSelectionContainerDelegate {
+  _SelectableTextContainerDelegate(
+    GlobalKey textKey,
+  ) : _textKey = textKey;
+
+  final GlobalKey _textKey;
+  RenderParagraph get paragraph => _textKey.currentContext!.findRenderObject()! as RenderParagraph;
+
+  @override
+  SelectionResult handleSelectParagraph(SelectParagraphSelectionEvent event) {
+    final SelectionResult result = _handleSelectParagraph(event);
+    if (currentSelectionStartIndex != -1) {
+      _hasReceivedStartEvent.add(selectables[currentSelectionStartIndex]);
+    }
+    if (currentSelectionEndIndex != -1) {
+      _hasReceivedEndEvent.add(selectables[currentSelectionEndIndex]);
+    }
+    _updateLastEdgeEventsFromGeometries();
+    return result;
+  }
+
+  SelectionResult _handleSelectParagraph(SelectParagraphSelectionEvent event) {
+    if (event.absorb) {
+      final SelectionEvent synthesizedEvent = SelectParagraphSelectionEvent(globalPosition: event.globalPosition, absorb: true);
+      for (int index = 0; index < selectables.length; index += 1) {
+        dispatchSelectionEventToChild(selectables[index], synthesizedEvent);
+      }
+      currentSelectionStartIndex = 0;
+      currentSelectionEndIndex = selectables.length - 1;
+      return SelectionResult.next;
+    }
+
+    // First pass, if the position is on a placeholder then dispatch the selection
+    // event to the [Selectable] at the location and terminate.
+    for (int index = 0; index < selectables.length; index += 1) {
+      bool globalRectsContainsPosition = false;
+      if (selectables[index].boundingBoxes.isNotEmpty) {
+        for (final Rect rect in selectables[index].boundingBoxes) {
```

Comment:
This method could be a performance bottleneck in a large document, but currently that's true of a lot of places, so probably not a big deal.

Reply:
Yeah, i'm also not the biggest fan of this, but couldn't come up with a better way to do this logic. What I might be able to do is instead of checking every selectable, I can check only the placeholders which are guaranteed to only have one item in their `boundingBoxes` list because they are `SelectionContainer`s https://github.com/flutter/flutter/blob/7482962148e8d758338d8a28f589f317e1e42ba4/packages/flutter/lib/src/widgets/selection_container.dart#L203-L204.

Reply:
Adding onto my last comment, a placeholder is only guaranteed to have one item in their `boundingBoxes` list if it is a `SelectionContainer` which is only guaranteed when the placeholder is a `Text` widget itself.